x-pack/plugins/lens/public/app_plugin/app.tsx

```diff
@@ -205,150 +208,38 @@ export function App({
-            overlays,
-          }
-        );
-      } catch (e) {
-        // ignore duplicate title failure, user notified in save modal
-        return;
-      }
-    }
-    try {
-      const newInput = (await attributeService.wrapAttributes(
-        docToSave,
-        options.saveToLibrary,
-        originalInput
-      )) as LensEmbeddableInput;
-
-      if (saveProps.returnToOrigin && redirectToOrigin) {
-        // disabling the validation on app leave because the document has been saved.
-        onAppLeave((actions) => {
-          return actions.default();
-        });
-        redirectToOrigin({ input: newInput, isCopied: saveProps.newCopyOnSave });
-        return;
-      } else if (saveProps.dashboardId && redirectToDashboard) {
-        // disabling the validation on app leave because the document has been saved.
-        onAppLeave((actions) => {
-          return actions.default();
-        });
-        redirectToDashboard(newInput, saveProps.dashboardId);
-        return;
-      }
-
-      notifications.toasts.addSuccess(
-        i18n.translate('xpack.lens.app.saveVisualization.successNotificationText', {
-          defaultMessage: `Saved '{visTitle}'`,
-          values: {
-            visTitle: docToSave.title,
-          },
-        })
-      );
-
-      if (
-        attributeService.inputIsRefType(newInput) &&
-        newInput.savedObjectId !== originalSavedObjectId
-      ) {
-        chrome.recentlyAccessed.add(
-          getFullPath(newInput.savedObjectId),
-          docToSave.title,
-          newInput.savedObjectId
-        );
-
-        dispatchSetState({ isLinkedToOriginatingApp: false });
-
+  const runSave = (saveProps: SaveProps, options: { saveToLibrary: boolean }) => {
+    return runSaveLensVisualization(
+      {
+        lastKnownDoc,
+        getIsByValueMode,
+        savedObjectsTagging,
+        initialInput,
+        redirectToDashboard,
+        redirectToOrigin,
+        setIsSaveModalVisible: () => {
+          setIsSaveModalVisible(false);
+        },
+        persistedDoc: appState.persistedDoc,
+        onAppLeave,
+        redirectTo,
+        ...lensAppServices,
+      },
+      saveProps,
+      options
+    ).then(
+      (newState) => {
+        dispatchSetState(newState);
         setIsSaveModalVisible(false);
-        // remove editor state so the connection is still broken after reload
-        stateTransfer.clearEditorState(APP_ID);
-
-        redirectTo(newInput.savedObjectId);
-        return;
-      }
-
-      const newDoc = {
-        ...docToSave,
-        ...newInput,
-      };
-
-      dispatchSetState({
-        isLinkedToOriginatingApp: false,
-        persistedDoc: newDoc,
-        lastKnownDoc: newDoc,
-      });
-
-      setIsSaveModalVisible(false);
-    } catch (e) {
-      // eslint-disable-next-line no-console
-      console.dir(e);
-      trackUiEvent('save_failed');
-      setIsSaveModalVisible(false);
-    }
+      },
+      () => {}
```

Comment:
Perhaps some error handling here with toasts?

Reply:
actually it already has an error handling inside the `runSaveLensVisualization` method and also a toast, this is why i didn't want to duplicate that. I think perhaps, i can add a comment to explain that.